app/uk/gov/hmrc/gform/gform/processor/FormProcessor.scala

```diff
@@ -269,53 +270,81 @@ class FormProcessor(
-                implicit val b: BankAccountReputationConnector[Future] = bankAccountReputationConnector
-                DataRetrieveService[PersonalBankAccountExistence, Future]
-                  .retrieve(p, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-              case p: PersonalBankAccountExistenceWithName =>
-                val maybeRequestParams = DataRetrieve.requestParamsFromCache(cache.form, p.id)
-                implicit val b: BankAccountReputationConnector[Future] = bankAccountReputationConnector
-                DataRetrieveService[PersonalBankAccountExistenceWithName, Future]
-                  .retrieve(p, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-              case v: CompanyRegistrationNumber =>
-                val maybeRequestParams = DataRetrieve.requestParamsFromCache(cache.form, v.id)
-                implicit val b: CompanyInformationConnector[Future] = companyInformationConnector
-                DataRetrieveService[CompanyRegistrationNumber, Future]
-                  .retrieve(v, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-              case n: NinoInsights =>
-                val maybeRequestParams = DataRetrieve.requestParamsFromCache(cache.form, n.id)
-                implicit val ni: NinoInsightsConnector[Future] = ninoInsightsConnector
-                DataRetrieveService[NinoInsights, Future]
-                  .retrieve(n, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-              case a: BankAccountInsights =>
-                val maybeRequestParams = DataRetrieve.requestParamsFromCache(cache.form, a.id)
-                implicit val ai: BankAccountInsightsConnector[Future] = bankAccountInsightConnector
-                DataRetrieveService[BankAccountInsights, Future]
-                  .retrieve(a, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-              case e: Employments =>
-                val maybeRequestParams = DataRetrieve.requestParamsFromCache(cache.form, e.id)
-                implicit val c: GformConnector = gformConnector
-                DataRetrieveService[Employments, Future]
-                  .retrieve(e, processData.formModelOptics.formModelVisibilityOptics, maybeRequestParams)
-            }
-          }(_ => Option.empty.pure[Future])(_ => Option.empty.pure[Future])
-        } else Option.empty.pure[Future]
+          val initialVisibilityOptics = processData.formModelOptics.formModelVisibilityOptics
+          pageModel
+            .fold { singleton =>
+              singleton.page
+                .dataRetrieves()
+                .foldLeft(Future.successful(List.empty[DataRetrieveResult] -> initialVisibilityOptics)) {
+                  case (acc, r) =>
+                    acc.flatMap {
+                      case (results, optics) if r.`if`.map(optics.evalIncludeIfExpr(_, None)).getOrElse(true) =>
```

Comment:
Does this handle multiple transitively dependent DataRetrievals? I'm testing chain of 3 data retrievals and it evaluates them, even though theirs `if` should evaluate to `false`.

Reply:
It should handle, will have a look